ap/columnar_util.py

```diff
@@ -879,56 +938,57 @@ def run_requires(self, task: law.Task, reqs: Optional[dict] = None):
 
         # run this instance's requires function
         if self.requires_func:
-            self.requires_func(task, reqs)
+            self.requires_func(self, task, reqs)
 
         return reqs
 
     def setup(self, func: Callable[[law.Task, dict, dict], dict]) -> None:
         """
         Decorator to wrap a function *func* that should be registered as :py:attr:`setup_func`
-        which is used to perform a custom setup of objects. The function should accept three
+        which is used to perform a custom setup of objects. The function should accept four
         arguments,
 
-           - *task*, the :py:class:`law.Task` instance,
-           - *inputs*, a dictionary with input targets corresponding to the requirements created by
-             :py:meth:`run_requires`, and
-           - *producer_kwargs*, a dictionary into which arguments should be inserted that are later
-             on passed to the wrapped function.
+            - *self* (positional), the instance of the array function itself,
+            - *task*, the :py:class:`law.Task` instance,
+            - *inputs*, a dictionary with input targets corresponding to the requirements created by
+              :py:meth:`run_requires`, and
+            - *call_kwargs*, a dictionary into which arguments should be inserted that are later
+              on passed to the wrapped function.
 
         The decorator does not return the wrapped function.
         """
         self.setup_func = func
 
-    def run_setup(self, task: law.Task, inputs: dict, producer_kwargs: Optional[dict] = None):
+    def run_setup(self, task: law.Task, inputs: dict, call_kwargs: Optional[dict] = None) -> None:
         """
         Recursively runs the setup function of this and all other known instances (via
         :py:attr:`uses` and :py:attr:`produces`) given the *task* and *inputs* corresponding to the
-        requirements created by :py:func:`run_requires`. *producer_kwargs* defaults to an empty
+        requirements created by :py:func:`run_requires`. *call_kwargs* defaults to an empty
         dictionary which should be filled to store arguments which are later on passed to the
         wrapped function.
         """
-        if producer_kwargs is None:
-            producer_kwargs = {}
+        if call_kwargs is None:
+            call_kwargs = {}
 
         # run the setup of all other known instances
         for obj in self.uses | self.produces:
```

Comment:
Also run over `self.shifts` (see comment for `run_update`)

Reply:
True! I'll address this in the next commit.

Reply:
Thanks! Looks good now.